tensorflow/core/grappler/optimizers/auto_mixed_precision.cc

```diff
@@ -0,0 +1,1720 @@
+  std::set<string> fp16_graylist_;
+  std::set<string> fp16_clearlist_;
+  absl::flat_hash_set<const NodeDef*> should_process_nodes_;
+};
+
+bool AutoMixedPrecisionImpl::NodeHasFP16KernelForTypeAttr(
+    const NodeDef& node, TypeAttrId taid) const {
+  NodeDef node_copy(node);
+  if (node.device().empty()) {
+    string device_name = virtual_placer_.get_canonical_device_name(node);
+    node_copy.set_device(device_name);
+  }
+  if (!SetDataType(&node_copy, taid, DataType::DT_HALF)) {
+    return false;
+  }
+  return IsKernelRegisteredForNode(node_copy).ok();
+}
+
+Status AutoMixedPrecisionImpl::PrintDebugLogs(bool preop, size_t timestamp) {
+  string prepend_path;
+  TF_RETURN_IF_ERROR(ReadStringFromEnvVar(
+      "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_LOG_PATH", "", &prepend_path));
+  if (prepend_path == "") return Status::OK();
+
+  string suffix =
+      strings::StrCat("_", preop ? "preop" : kSuffix, "_", id_, "_", timestamp);
+
+  string fname =
+      io::JoinPath(prepend_path, strings::StrCat("graphdef", suffix, ".pb"));
+  std::fstream f;
+  f.open(fname.c_str(), std::fstream::out | std::fstream::binary);
+  f << graph_->SerializeAsString();
+  f.close();
+  LOG(INFO) << "Saved " << (preop ? "pre-optimization" : "post-optimization")
+            << " graph as binary to " << fname;
+
+  fname = io::JoinPath(prepend_path,
+                       strings::StrCat("graphdef", suffix, ".pb.txt"));
+  f.open(fname.c_str(), std::fstream::out);
+  f << graph_->DebugString();
+  f.close();
+  LOG(INFO) << "Saved " << (preop ? "pre-optimization" : "post-optimization")
+            << " graph as text to " << fname;
+
+  if (!preop) {
+    fname = io::JoinPath(prepend_path,
+                         strings::StrCat("paintbuckets", suffix, ".txt"));
+    f.open(fname.c_str(), std::fstream::out);
+    f << "WhiteList:\n";
+    for (auto x : AutoMixedPrecisionLists::WhiteList()) {
+      f << x << "\n";
+    }
+    f << "\nBlackList:\n";
+    for (auto x : AutoMixedPrecisionLists::BlackList()) {
+      f << x << "\n";
+    }
+    f << "\nGrayList:\n";
+    for (auto x : AutoMixedPrecisionLists::GrayList()) {
+      f << x << "\n";
+    }
+    f << "\nClearList:\n";
+    for (auto x : AutoMixedPrecisionLists::ClearList()) {
+      f << x << "\n";
+    }
+    f.close();
+    LOG(INFO) << "Saved paint bucket info to " << fname;
+  }
+  return Status::OK();
+}
+
+void AutoMixedPrecisionImpl::LogSkippedNode(const NodeDef& node) const {
+  VLOG(2) << "Skipping " << node.op() << " node " << node.name()
+          << " because it is "
+          << (MustPreserve(node)
+                  ? "on the fetch list"
+                  : "not on the GPU, or the GPU arch is not suitable");
+}
+
+bool AutoMixedPrecisionImpl::MustPreserve(const NodeDef& node) const {
+  return nodes_to_preserve_.count(node.name());
+}
+
+bool AutoMixedPrecisionImpl::IsOnGPU(const NodeDef& node) const {
+  string device_name;
+  if (node.device().empty()) {
```

Comment:
I'm not sure I understand this one, could you clarify? The identical code is used in layout_optimizer.cc.

Reply:
If layout_optimizer.cc does this as well, it is fine.

The point I was getting at was a user could do something like:

```
with tf.device('/replica:0'):
  x = <white-list op>
```

The optimizer would not touch `x`, despite it being on the GPU, because the string "GPU" does not appear in the device, and the device is not empty.